Fix soundness bug in SdHoareTripleChecker

maul-esel opened the pull request:
This change fixes an unsoundness bug in `SdHoareTripleChecker`, shown by the failing test `HoareTripleCheckerTests::sdHtcTest02`.

(It's probably best to review the two commits separately.)

maul-esel committed:
some code cleanup

Simplify code, and update some comments.
The functionality should be unchanged.

maul-esel committed:
fix unsoundness for Hoare triples with program constants

This change fixes the failing test in HoareTripleCheckerTests.

maul-esel committed:
add new failing test

maul-esel committed:
fix unsoundness in SdHoareTripleChecker (for postcondition false)

maul-esel committed:
fix method name typo

maul-esel commented:
After discussion with @Heizmann, we have decided to spend a little more time on this and make sure all soundness issues surrounding constants in the `SdHoareTripleChecker` are fixed (e.g. also for call / return), and to add missing support for axioms.

maul-esel committed:
Merge branch 'dev' into wip/dk/const-hoare-fix

maul-esel committed:
add abstract test suite for Hoare triple checkers

maul-esel committed:
extract nested classes from SdHoareTripleChecker

maul-esel committed:
add statistics to helper classes

maul-esel committed:
copy internal Hoare triple checks into InternalCheckHelper

maul-esel committed:
Hoare triples with infeasible calls or returns are also valid

maul-esel committed:
remove unnecessary null checks

The SdHoareTripleChecker always provides an IPredicateCoverageChecker.
It does not make sense to sometimes check if it is null, and other times
assume it is not.

maul-esel committed:
SmtParserUtils: add support for old vars

maul-esel committed:
add new tests involving old vars

Inspiration taken from commit 41f53ffe7c.

maul-esel committed:
fix tests relating to old vars

maul-esel committed:
also perform tests on MonolithicHoareTripleChecker

maul-esel committed:
Merge branch 'dev' into wip/dk/const-hoare-fix

schuessf commented:
Compared the latest nightly build [#53](https://jenkins.sopranium.de/job/Ultimate/job/Ultimate%20Nightly/job/PR-591/53/) with the reference build [#302](https://jenkins.sopranium.de/job/Ultimate/job/Ultimate%20Nightly/job/dev/302/) of the reference branch.

The following tests failed:
* I_programs_regression_c_NutzTransformation02.c S_programs_regression_c_BlockEncodingV2AutomizerC-FP-MaxSaneBE.epf T_programs_regression_c_BlockEncodingV2AutomizerC.xml (de.uni_freiburg.informatik.ultimate.regressiontest.generic.RegressionTestSuite)
* I_programs_regression_c_TestIntegerDivision01.c S_programs_regression_c_AutomizerC-BitvectorTranslation.epf T_programs_regression_c_AutomizerC.xml (de.uni_freiburg.informatik.ultimate.regressiontest.generic.RegressionTestSuite)

Please check these tests before merging this PR.

maul-esel committed:
some simplification in SdHoareTripleCheckHelper::check

maul-esel committed:
fix typo

maul-esel committed:
simplify InternalCheckHelper::isInductiveSelfloop

maul-esel committed:
code cleanup and improvements for InternalCheckHelper

maul-esel committed:
more Hoare triple checker tests

maul-esel committed:
move code for dataflow checks of call Hoare triples to CallCheckHelper

maul-esel committed:
remove incorrect comment

maul-esel committed:
add some Hoare triple tests for calls

maul-esel committed:
more tests involving constants

maul-esel committed:
remove obsolete debugging code

- A similar effect can be achieved through the "review" functionality
  in ChainingHoareTripleChecker.
- We also have unit tests for IHoareTripleCheckers now

maul-esel committed:
delete obsolete methods (but keep informative comment)

maul-esel committed:
move code for return checks to ReturnCheckHelper

maul-esel committed:
minor code improvement in test suite

maul-esel committed:
some comments and slight improvement in CallCheckHelper, more tests

maul-esel committed:
make SdHoareTripleCheckHelper generic

maul-esel committed:
more tests and fixes for Hoare triples with unmodifiable old vars

maul-esel committed:
simplify code in CallCheckHelper

maul-esel committed:
extend doc comment

maul-esel committed:
rename field

maul-esel committed:
fix tests

maul-esel commented:
In its current form, the PR

* fixes a known bug with constants in internal statements
* refactors and simplifies the code (though more could be done, especially for calls and returns, in the future)
* adds a number of additional unit tests for Hoare triple checkers, including one test that reveals an existing bug in the handling of calls

Things left to do in the future:

* further refactor and simplify the code
* document the assumptions made about call and return statements, including in the case of interprocedural compositions
* make fewer assumptions
* implement `IAbstractPredicate::getConsts` and fix the bug with calls
* ensure support for function symbols and axioms

As discussed, I'll merge this PR now.
